File: 2326-spiral-matrix-iv/2326-spiral-matrix-iv.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func spiralMatrix(m int, n int, head *ListNode) [][]int {

	answer := make([][]int, m)
	for i := 0; i < m; i++ {
		answer[i] = make([]int, n)
		for j := 0; j < n; j++ {
			answer[i][j] = -1
		}
	}

	direction := 0
	minRow := 1
	minCol := 0
	maxRow := m - 1
	maxCol := n - 1
	row := 0
	col := 0

	var update func(int)
	update = func(i int) {
        answer[row][col] = i
        
		switch direction {
		case 0:
			col++;
			if col > maxCol {
				col--
				row++
			} else if col == maxCol {
				maxCol--
				direction = 1
			}
		case 1:
			row++
			if row == maxRow {
				maxRow--
				direction = 2
			}
		case 2:
			col--
			if col == minCol {
				minCol++
				direction = 3
			}
		case 3:
			row--
			if row == minRow {
				minRow++
				direction = 0
			}
		}
	}

	for {
		if head == nil {
			return answer
		}
		update(head.Val)
		head = head.Next
	}
}

File: 2326-spiral-matrix-iv/2326-spiral-matrix-iv.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
    class Solution {

        private int[][] array;
        private int row = 0;
        private int col = 0;
        private int minRow = 1;
        private int maxRow;
        private int minCol = 0;
        private int maxCol;
        private int direction = 0;


        public int[][] spiralMatrix(int m, int n, ListNode head) {
            array = new int[m][n];
            maxRow = m-1;
            maxCol = n-1;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    array[i][j] = -1;
                }
            }

            ListNode node = head;
            while (node != null) {
                update(node.val);
                node = node.next;
            }
            return array;
        }

        private void update(int element) {
            array[row][col] = element;

            switch (direction) {
                case 0:
                    col++;
                    if (col > maxCol) {
                        col--;
                        row++;
                    } else if (col == maxCol) {
                        maxCol--;
                        direction = 1;
                    }
                    break;
                case 1:
                    row++;
                    if (row == maxRow) {
                        maxRow--;
                        direction = 2;
                    }
                    break;
                case 2:
                    col--;
                    if (col == minCol) {
                        minCol++;
                        direction = 3;
                    }
                    break;
                case 3:
                    row--;
                    if (row == minRow) {
                        minRow++;
                        direction = 0;
                    }
                    break;
            }
        }
    }